kong/pdk/response.lua

```diff
@@ -633,129 +633,199 @@ local function new(self, major_version)
-      error("code must be a number", 2)
+      if body ~= nil and type(body) ~= "string" and type(body) ~= "table" then
+        error("body must be a nil, string or table", 2)
+      end
 
-    elseif status < MIN_STATUS_CODE or status > MAX_STATUS_CODE then
-      error(fmt("code must be a number between %u and %u", MIN_STATUS_CODE, MAX_STATUS_CODE), 2)
-    end
+      if headers ~= nil and type(headers) ~= "table" then
+        error("headers must be a nil or table", 2)
+      end
 
-    if body ~= nil and type(body) ~= "string" and type(body) ~= "table" then
-      error("body must be a nil, string or table", 2)
-    end
+      if headers ~= nil then
+        validate_headers(headers)
+      end
 
-    if headers ~= nil and type(headers) ~= "table" then
-      error("headers must be a nil or table", 2)
-    end
+      local ctx = ngx.ctx
 
-    if headers ~= nil then
-      validate_headers(headers)
-    end
+      if is_buffered_exit then
+        self.ctx.core.buffered_status = status
+        self.ctx.core.buffered_headers = headers
+        self.ctx.core.buffered_body = body
 
-    local ctx = ngx.ctx
+      else
+        ctx.KONG_EXITED = true
+      end
 
-    if is_buffered_exit then
-      self.ctx.core.buffered_status = status
-      self.ctx.core.buffered_headers = headers
-      self.ctx.core.buffered_body = body
+      if ctx.delay_response and not ctx.delayed_response then
+        ctx.delayed_response = {
+          status_code = status,
+          content     = body,
+          headers     = headers,
+        }
 
-    else
-      ctx.KONG_EXITED = true
+        ctx.delayed_response_callback = flush
+        coroutine.yield()
+
+      else
+        return send(status, body, headers)
+      end
     end
 
-    if ctx.delay_response and not ctx.delayed_response then
-      ctx.delayed_response = {
-        status_code = status,
-        content     = body,
-        headers     = headers,
-      }
+  else
+    local VALID_CODES = {
+      [200] = true,
+      [400] = true,
+      [403] = true,
+      [500] = true,
+      [502] = true,
+      [503] = true,
+      -- NOTE: when adding new code, change the documentation and error
+      -- message raised below accordingly
+      --
+      -- Code are from http://lxr.nginx.org/source/src/stream/ngx_stream.h#0029
+    }
+
+    function _RESPONSE.exit(status, body, headers)
+      if type(status) ~= "number" then
+        error("code must be a number", 2)
+
+      elseif not VALID_CODES[status] then
+        error("unacceptable code, only 200, 400, 403, 500, 502 and 503 " ..
+              "are accepted", 2)
+      end
 
-      ctx.delayed_response_callback = flush
-      coroutine.yield()
+      if body ~= nil and type(body) ~= "string" then
+        error("body must be a nil or a string", 2)
+      end
 
-    else
-      return send(status, body, headers)
+      if body then
+        if status < 500 then
+          local res, err =  ngx.print(body)
```

Comment:
This is the only thing that I am wondering if it is worth it. It means that plugin probably has to understand the underlying protocol, and we speak about raw TCP which makes it really hard for plugins to deal with the output (especially when multiple different TCP protocols are used with the stream). It is good that there is this `< 500` at least, which probably covers a lot already. What if we make it `< 400`, and leave `200` as the only means for plugins to send data back to client.